src/pip/_internal/index.py

```diff
@@ -487,18 +502,52 @@ def _sort_key(self, candidate):
                 build_tag = (int(build_tag_groups[0]), build_tag_groups[1])
         else:  # sdist
             pri = -(support_num)
-        return (binary_preference, candidate.version, build_tag, pri)
+        yank_value = -1 * int(link.is_yanked)  # -1 for yanked.
+        return (
+            yank_value, binary_preference, candidate.version, build_tag, pri,
+        )
 
-    def get_best_candidate(self, candidates):
-        # type: (List[InstallationCandidate]) -> InstallationCandidate
+    # Don't include an allow_yanked default value to make sure each call
+    # site considers whether yanked releases are allowed. This also causes
+    # that decision to be made explicit in the calling code, which helps
+    # people when reading the code.
+    def get_best_candidate(
+        self,
+        candidates,    # type: List[InstallationCandidate]
+        allow_yanked,  # type: bool
+    ):
+        # type: (...) -> Optional[InstallationCandidate]
         """
         Return the best candidate per the instance's sort order, or None if
-        no candidates are given.
+        no candidate is acceptable.
+
+        :param allow_yanked: Whether to permit returning a yanked candidate
+            in the sense of PEP 592. If true, a yanked candidate will be
+            returned only if all candidates have been yanked.
         """
         if not candidates:
             return None
 
-        return max(candidates, key=self._sort_key)
+        best_candidate = max(candidates, key=self._sort_key)
+
+        # Log a warning per PEP 592 if necessary before returning.
+        link = best_candidate.location
+        if not link.is_yanked:
+            return best_candidate
+
+        # Otherwise, all the candidates were yanked.
+        if not allow_yanked:
```

Comment:
nit: I do feel it'd be better to (if allow_yanked is False) prevent yanked releases from ever reaching `CandidateEvaluator`, by filtering in iter_applicable; since yanked releases are not applicable if allow_yanked is False.

None the less, I don't want to block on this and we can come around to this later too.

Reply:
Oh, I wouldn't consider this a nit. I think it's an important point! (And sorry I missed this before. I see now you were making two suggestions in the same paragraph, and I thought it was just one.) I think it's a good suggestion and agree we should do it, but I wanted to point out a couple nuances.

First, it seems like this would split the handling not just into two different places, but three. This is because `pip list` doesn't use `iter_applicable()` but does its own pre-filtering: https://github.com/pypa/pip/blob/0dbab23df903926de71b8e2ce08359a378327ad1/src/pip/_internal/commands/list.py#L181-L186

So `pip list` would also have to know how to filter out yanked candidates. I think the right solution here might be to have that `pip list` filtering logic go through `iter_applicable()` instead somehow (perhaps generalized slightly if needed). That would let us keep that logic in one place.

Second, it seems like there's another option to filtering in `iter_applicable()`, and that's to filter at the `Link` level before an `InstallationCandidate` is even created. This is done in the `CandidateEvaluator` class by the `CandidateEvaluator.evaluate_link()` method: https://github.com/pypa/pip/blob/0dbab23df903926de71b8e2ce08359a378327ad1/src/pip/_internal/index.py#L345-L348

And here is the call site (in `get_install_candidate()`): https://github.com/pypa/pip/blob/0dbab23df903926de71b8e2ce08359a378327ad1/src/pip/_internal/index.py#L1029-L1035

If we're able to filter things out earlier like that, it seems like it would be better to do it sooner (e.g. since it doesn't seem like there'd be any harm). I believe this would be another way to solve the `pip list` issue, too.

Reply:
Yea, I only called it a nit because I didn't mean to block this PR on this -- it's probably something worth tackling in a separate PR. :)

I'd say let's file an issue for this change, so that we have something to track/discuss fixing this stuff, after this PR.